llvm/include/llvm/IR/IntrinsicsAMDGPU.td

```diff
@@ -2819,11 +2819,11 @@ def int_amdgcn_fdot2_f16_f16 :
 def int_amdgcn_fdot2_bf16_bf16 :
   ClangBuiltin<"__builtin_amdgcn_fdot2_bf16_bf16">,
   DefaultAttrsIntrinsic<
-    [llvm_i16_ty],   // %r
+    [llvm_bfloat_ty],   // %r
```

Comment:
This is the change in the existing interface, it will break programs already using it. It also will break clang builtin.

You can get away with changing only instruction's profile to use BF16 and tweak pattern with casts to bf16.

Reply:
It will just maybe require changing the clang builtin codegen to insert the casts. You still need to make this change

Reply:
You should also add IR auto upgrade

Reply:
Does it make sense if we also update both the Clang builtins and LLVM builtins here, and use IR auto upgrade to rewrite existing uses?

Reply:
Changing the clang builtin breaks user code. Anyone using it will now have an implicit itofp cast to bfloat. For the first step, I would leave the clang builtin alone (and only make the codegen change to emit the new IR). We would either need to introduce a new builtin with the corrected signature, or add a language flag to change the behavior

Reply:
clang/test/CodeGenOpenCL/builtins-amdgcn-dl-insts-gfx11.cl fails. You need to insert casts to bf16 while lowering it to make it working.

Reply:
The cast will be inserted automatically in `clang/lib/CodeGen/CGBuiltin.cpp` after removing the two assertions. I reverted my change to the test case by accident. Lol